### PR TITLE
Methods for Checks API

### PR DESCRIPTION
**Pull request by axel-op**
This is a non-breaking PR. 

I have implemented some methods to interact with the Checks API (https://developer.github.com/v3/checks) for my personal use, I thought I could make this draft PR to merge them later.

### CODE REVIEW

**Comment by robrbecker** [comment]
Thanks! Checks API would be fantastic to have!

**Comment by axel-op** [comment]
@robrbecker I don't think I'll progress on this anytime soon. 
I have implemented some methods in https://developer.github.com/v3/checks/runs.
Remain the methods in https://developer.github.com/v3/checks/suites.

Also, I have not implemented `listCheckRunsForRef` and `listCheckRunsInSuite` as their response format is not a list as it is usually the case with multi-pages responses, and I don't think that currently internal methods can handle this format to return a stream.

**Comment by robrbecker** [comment]
Thanks for your work. I'm not in a hurry if you want to take your time. Or maybe you're saying you're not planning on finishing it and you'd like someone to take it over?

**Comment by axel-op** [comment]
> Or maybe you're saying you're not planning on finishing it and you'd like someone to take it over?

This is what I meant. I don't need the unimplemented methods and I have other priorities for now.

**Comment by axel-op** [comment]
@robrbecker While merging your recent changes into my branch, I had to make the [`requestJson` method](https://github.com/SpinlockLabs/github.dart/pull/182/files#diff-5090c09fd799b45e60bf2bed9338e354R201) public. Tell me if this is OK for you.

**Comment by robrbecker** [comment]
everything else is public, so I think that'll be fine.

**Comment by axel-op** [comment]
I've finally implemented the methods for every endpoint of the Checks API.

This PR is ready for review.

File: lib/src/util.dart [hunk 1]
```diff
@@ -15,7 +17,8 @@ String buildQueryString(Map<String, dynamic> params) {
     if (params[key] == null) {
       continue;
     }
-    queryString.write('$key=${Uri.encodeComponent(params[key].toString())}');
+    // TODO: change to library-specific json encoding
+    queryString.write('$key=${Uri.encodeComponent(jsonEncode(params[key]))}');
```
Comment (robrbecker) [inline lib/src/util.dart:21]: Can you explain why we're `jsonEncode`ing all objects that are values? This makes me a little nervous that we might break other things so I'd like to understand it better? 🤷 

> change to library-specific json encoding
Also, I'm not sure I know what you mean by that. Could you give me more information so when I or another person wants to address this TODO we know what we're doing?

File: lib/src/common/model/checks.dart [hunk 2]
```diff
@@ -0,0 +1,407 @@
+import 'package:github/src/util.dart';
+import 'package:meta/meta.dart';
+
+@immutable
+abstract class _EnumWithValue {
```
Comment (robrbecker) [inline lib/src/common/model/checks.dart:5]: This seems like a fairly general utility that could be reused. Would you be opposed to moving it to utils.dart and making a public class?

File: lib/src/common/github.dart [hunk 3]
```diff
@@ -79,67 +80,53 @@ class GitHub {
...
-    return _search ??= SearchService(this);
-  }
+  SearchService get search => _search ??= SearchService(this);
 
   /// Service to provide a handy method to access GitHub's url shortener.
-  UrlShortenerService get urlShortener {
-    return _urlShortener ??= UrlShortenerService(this);
-  }
+  UrlShortenerService get urlShortener =>
+      _urlShortener ??= UrlShortenerService(this);
```
Comment (robrbecker) [inline lib/src/common/github.dart:121]: Thanks for shortening up this code and above.

**Comment by robrbecker** [comment]
@axel-op  This looks really good. Just 2 questions really and you'll need to resolve conflicts with master (a method was added in master)

File: lib/src/util.dart (hunk 1 above)
Comment (axel-op) [reply lib/src/util.dart:21]: You're right. I made this change following the edit of the type of accepted values in this `params` map (`String` to `dynamic`), but on second thought it's a little risky.

This TODO was related to my other PR, I'll remove it.

My idea was to call an encoding function to serialize values that wouldn't be strings.

File: lib/src/common/model/checks.dart [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    if (other is CheckRunAnnotation) {
+      return other.annotationLevel == this.annotationLevel &&
+          other.path == this.path &&
+          other.startColumn == this.startColumn &&
+          other.endColumn == this.endColumn &&
+          other.startLine == this.startLine &&
+          other.endLine == this.endLine &&
+          other.title == this.title &&
+          other.message == this.message &&
+          other.rawDetails == this.rawDetails;
```
Comment (robrbecker) [inline lib/src/common/model/checks.dart:240]: I'm getting some lints that all these `this.` are redundant and can be removed

**Pull request merged**